[INSTALL] Unify cmake install functions and dynamically set component dependencies

dbarker opened the pull request:
Fixes #3293 as followup to #3220

## Changes

- Unifies the cmake install commands into functions 
   - `otel_add_component()` 
      -  Create a component with targets and optional files for install
   - `otel_install_components()`
      - Install all components created 
   - `otel_install_cmake_config()` 
      - Installs the opentelemetry-cpp-config.cmake` and version files. 
   - `otel_install_thirdparty_definitions()`
      - Installs the third-party definitions file (cmake search modes and built-with versions) 
- Removes the static component definitions and mappings 
   -  component to target, component to component, and component to thirdparty dependency mappings are now determined dynamically based on the CMake target properties [`LINK_LIBRARIES`](https://cmake.org/cmake/help/v4.0/prop_tgt/LINK_LIBRARIES.html) and [`INTERFACE_LINK_LIBRARIES`](https://cmake.org/cmake/help/v4.0/prop_tgt/INTERFACE_LINK_LIBRARIES.html).    
- Adds `opentelemetry-cpp::` alias targets to support install and unify the public cmake interface to `opentelemetry-cpp` 
  - The same `opentelemetry-cpp::<target>` names imported with `find_package` can now be used when `opentelmetry-cpp` is built directly as a submodule (`add_subdirectory`) in a larger project. 
  - Adds a test in ci to verify the alias targets are created for the api and sdk components when using CMake's FetchContent module.
  - Updates the INSTALL documentation to add notes on FetchContent usage 

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

dbarker commented:
Hi @owent, This PR is based on your requests for followup to #3220. Can you take a look and give some feedback?

owent commented:
> Hi @owent, This PR is based on your requests for followup to #3220. Can you take a look and give some feedback?

Sorry for the delay. I will review it some time later.

owent commented on cmake/otel-install-functions.cmake:
```diff
@@ -0,0 +1,504 @@
+  endforeach()
+
+  configure_file(
+    "${CMAKE_SOURCE_DIR}/cmake/templates/thirdparty-dependency-definitions.cmake.in"
```
Maybe it's better to use the simular path as below? (`"${CMAKE_CURRENT_LIST_DIR}/templates/thirdparty-dependency-definitions.cmake.in"`)

dbarker replied:
Good catch. You're change is right. It should use `CMAKE_CURRENT_LIST_DIR` to ensure the project builds/installs correctly if used as a submodule.

dbarker replied:
I've pushed a comment to use PROJECT_SOURCE_DIR here and where we set/get properties used for install. That should work better when this project is used as a submodule/subdirectory in a larger project.

owent commented on cmake/find-package-support-functions.cmake:
```diff
@@ -190,23 +131,16 @@ endfunction()
     set(is_required FALSE)
-    is_dependency_required(${dependency} ${components_in} is_required)
-    message(DEBUG "find_required_dependencies: dependency = ${dependency}, is_required = ${is_required}")
+    is_dependency_required(${_dependency} ${components_in} is_required)
```
Can we cache the result if a dependency is already imported. There are too many recursive loops here.

dbarker replied:
Done

dbarker commented on cmake/otel-install-functions.cmake:
```diff
@@ -0,0 +1,504 @@
+  # Write config file for find_package(opentelemetry-cpp CONFIG)
+  set(INCLUDE_INSTALL_DIR "${CMAKE_INSTALL_INCLUDEDIR}")
+  configure_package_config_file(
+    "${CMAKE_CURRENT_LIST_DIR}/cmake/templates/opentelemetry-cpp-config.cmake.in"
```
> Should it be "${CMAKE_CURRENT_LIST_DIR}/templates/opentelemetry-cpp-config.cmake.in" ?

No, since  the `cmake/templates` directory with the `.in` file is in the root of the repo we need to keep the `cmake` folder in the path of this arg.  Line 485 works now because the `otel_install_cmake_config` function is called from the top level cmake file of the repo.

dbarker replied:
> Should we create the directory ${CMAKE_CURRENT_BINARY_DIR}/cmake/${PROJECT_NAME} before generate config file into it?

I lean towards no. CMake's `configure_file` creates the directory if it doesn't exist.

owent commented:
I'm on holiday these days.Can I continue to review several day's later

dbarker commented:
@owent - for when you get back from holiday  :)

I've added a new test using CMake's FetchContent module and updated INSTALL documentation to include this as a supported approach to incorporate `opentelemetry-cpp` as a sub-module/sub-directory in an external CMake project.

owent commented on cmake/templates/component-definitions.cmake.in:
```diff
@@ -0,0 +1,28 @@
+# COMPONENT to TARGET lists
+# ----------------------------------------------------------------------
+
+@OTEL_COMPONENTS_TARGETS_BLOCK@
```
Can we also generate the target list comments in `cmake/templates/opentelemetry-cpp-config.cmake.in`?

dbarker replied:
Yeah. That is a good idea. What do you think about just removing the current comments listing targets and components then adding a comment pointing to the `component-definitions.cmake` file installed? That file contains all the information - just not in an .rst format.

dbarker replied:
Added a follow up issue. #3398

owent reviewed (approved):
Excellent job. I have already tested in my project. It works well.

marcalff commented:
> Excellent job. I have already tested in my project. It works well.

This says it all. Ok to merge.

owent commented:
> > Excellent job. I have already tested in my project. It works well.
> 
> This says it all. Ok to merge.

Yes, LGTM. I think we can continue to work on generating document comments in another PR.

marcalff reviewed (approved):
LGTM, thanks for the fix.
